### DISCUSSION TITLE
print pdf

### DISCUSSION THREAD

**Discussion by persiabudi28** [dotnet/maui#13528] [General]
how to print pdf?

**Comment by Cywizz** [comment]
The PrintManager.GetForCurrentView() is not supported for WinUI desktop applications (Used in the link above). I am struggling to find a solution to print a pdf document in .Net Maui (Windows platform). 
Did anyone manage to get around the issue, or is printing just not supported at this stage?

**Reply by stipar** [reply]
yo logre imprimir en windows, lo que todavía no es posible es en android me parece

**Reply by LanceMcCarthy** [reply]
> yo logre imprimir en windows, lo que todavía no es posible es en android me parece

por favor mire el resto del código en mi demostración. Funciona para todas las plataformas.

el código específico de la plataforma está en una clase parcial, el nombre del archivo dice para qué plataforma es.

**Reply by kushald7** [reply]
@stipar  can you share code for printing on windows

**Comment by LanceMcCarthy** [comment]
In case it helps, I have a .NET MAUI demo that shows you how to do this using the native platform APIs to print a PDF file with a single line of code (after you've added the helper classes).

```csharp
private async void PrintPdf_OnClicked(object sender, EventArgs e)
{
    var stream = await FileSystem.OpenAppPackageFileAsync("pdfviewer-overview.pdf");

    // See the 3 classes in /Helpers/ folder.
    var helper = new PrintHelper();

#if WINDOWS
    
    helper.Print(stream, "pdfviewer-overview.pdf", this.Window.Handler.PlatformView as Microsoft.UI.Xaml.Window);

#elif __IOS__ || MACCATALYST || ANDROID

    helper.Print(stream, "pdfviewer-overview.pdf");
    
#endif
}
```


See the project for the rest of the info => https://github.com/LanceMcCarthy/CustomMauiExamples/tree/main/src/DocumentPrinting

**Reply by stipar** [reply]
estoy probando en .net8 no funciona, alguna solucion?

**Comment by Cywizz** [comment]
Thank you @LanceMcCarthy , I will check definitely this out!

**Comment by Xyncgas** [comment]
Anyone has success printing PDF by creating a webview with 0 width height and using javascript to do it

**Reply by LanceMcCarthy** [reply]
This wont work on all platforms, because not all the native WebViews support printing via JavaScript. 

You also have to consider figuring out how to get the PDF file from outside of the webview to inside of the webview without user interaction or messing up the PDF. Sure you can try some crazy conversations like serializing the byte[] to base64, then embedding the base64 into the dynamically loaded javascript code as a variable, but you can't do this for anything but trivial documents.

**Reply by Xyncgas** [reply]
>You also have to consider...

Remember it's Maui, I can share memory between webview and .Net

**Comment by martinblampied** [comment]
Anyone find a work around for this?

**Reply by LanceMcCarthy** [reply]
Not sure what you mean, did you see my demo above that shows how to accomplish printing? check it out here https://github.com/LanceMcCarthy/CustomMauiExamples/tree/main/src/DocumentPrinting

tl;dr: You use the native platform's printing APIs and everything is peachy (assuming it's a PDF file you want to print).

**Reply by martinblampied** [reply]
Ah sorry miss read this a bit. 

I have a MAUI windows only app that has a XAML screen with a load of input entries. I want to export that screen to a PDF. I was wondering if a 0 height webview with injected JS would work but don't think it would.

Any ideas?

**Reply by Cywizz** [reply]
You can use MAUI's screenshot capability if you just want a copy of what is on the screen. Look at the .CaptureAsync documentation.
I use QuestPDF in my MAUI app (all platforms) and it works well to create a pdf for you. It uses code-only to generate the pdf  and will allow you to take those data fields and make it pretty. It does support embedding images if you just want to dump the screenshot in a pdf.

After that, Lance's document printing implementation works great, I just tweaked mine to use the platform specific classes instead of platform directives in code.

**Reply by martinblampied** [reply]
Thanks your probably right thats the best way to do it. I was hoping (being lazy) that there was a way to export/print my current XAML directly to PDF without having the extra steps of "rewriting" the xaml into QuestPDF but I guess it give me more flexibility in the long run. Will take a look at screenshot but dont think it will give what I need.

**Reply by LanceMcCarthy** [reply]
I've found the screenshot option works excellent for this.

I have a demo that shows how to do this with both full screen or partial screen capture, and it even generates a PDF file. See the sub Readme for the project here  https://github.com/LanceMcCarthy/CustomXamarinDemos/tree/main/src/RenderImage

I haven't had time to convert it to maui yet, but there no difference in code.

**Reply by MichaelHall001** [reply]
I've downloaded your Printing Demo for .net MAUI.  It works on Windows but I need it to work on Android.  I have these squiggly red lines on line 17 and 25.  Also the errors at the bottom of Visual Studio.  Thoughts on this?  Should it work on Android? 
![error](https://github.com/dotnet/maui/assets/90070115/d55a5dee-245d-4e3e-9ecf-9d14f8161c47)

**Reply by LanceMcCarthy** [reply]
Heh, that's unexpected. I'm away from a PC for couple days, but I'll use Github Codespaces to debug it in a few hours to see if I can fix it sooner.

In the meantime, you can always copy/paste the code out of the Helper class file and into the code behind to use the Print() method directly.

I'll reply again once I've had a chance to review it.

**Reply by stipar** [reply]
el metodo print ya no esta disponible para android, si para las otras plataformas, seguramente no debe ser compatible con el nuevo sdk de android

**Reply by Cywizz** [reply]
I think Lance's demo is just missing a filename in the constructor if I remember correctly as he does not have a default constructor specified.
I have successfully used his code to get things going on Windows and Android (12) and have extracted it in a demo here: https://github.com/Cywizz/PrintDemo
I have yet to test it on other platforms, although the code is there.
All I changed was adding a Print Service for injection as I am using MVVM in my main project (Not this demo), and added a CanPrint() method where you can implement additional logic before the process starts. I also moved the "helper" classes into their platform specific locations. 
Thanks again for the initial code Lance, I have credited you in the demo.

**Reply by LanceMcCarthy** [reply]
Thanks @Cywizz! 

@MichaelHall001  I'm sorry for the delay, I was on vacation last week and GitHub CodeSpaces was not sufficient enough for me to say "yeah, it's fixed". I have updated the demo and used Cywizz's abstraction approach to surface the common naming to the MAUI layer:

![image](https://github.com/dotnet/maui/assets/3520532/2b7951aa-7c2d-4552-8edd-e605166a2444)

You can see the new code at the following locations:

- https://github.com/LanceMcCarthy/CustomMauiExamples/blob/main/src/DocumentPrinting/DocumentPrinting/MainPage.xaml.cs
- https://github.com/LanceMcCarthy/CustomMauiExamples/blob/main/src/DocumentPrinting/DocumentPrinting/Helpers/PrintHelper.Android.cs

**Comment by MichaelHall001** [comment]
Its working LikeABo$$ OMG!!! Thanks Lance!! You are a life saver.  You are
the best!!!

**Comment by MichaelHall001** [comment]
We never were able to get it working.  It came up with no error messages
when we compiled it after you fixed it.  My client couldn't get his blue
tooth printer to talk to the phone and stuff.  We abandoned everything.
Sorry.

**Reply by LanceMcCarthy** [reply]
It looks like Lucas found an answer for the FileNotFoundException and deleted the question.

If anyone else lands on this message and is wondering how to handle this, please follow this tutorial to understand how use various ways to work with files in a .NET MAUI application https://learn.microsoft.com/en-us/training/modules/dotnet-files/

**Reply by jaberlinz** [reply]
Hi McCarthy, I ran the project on windows and it works well.
But on android it throw this error
**System.IO.DirectoryNotFoundException:** 'Could not find a part of the path /data/user/0/com.companyname.mauiapp1/files/Documents/AboutAssets.txt'.'
The link you have provided does not suggest anything in relation to this behaviour.
Could you please advice further.
Thanks!

**Reply by LanceMcCarthy** [reply]
That isnt related to the demo, it means that you have a file that was supposed to be compiled with the app, but it wasn't. You can google that error and find conversations about how to handle it. Sometimes when I hit these, I'll start with a fresh project because all of the csproj stuff hasnt been borked.

Search something like  ".NET MAUI 'Could not find a part of the path AboutAssets.txt', but the best place to get help with this kind of compilation issue is in the .NET Discord (the MAUI channel) => http://aka.ms/dotnet-discord

**Reply by jaberlinz** [reply]
thanks for the swift answer!
I followed this link (https://stackoverflow.com/questions/72838980/system-io-directorynotfoundexception-could-not-find-a-part-of-the-path)
it sais: please be aware that Environment.SpecialFolder.LocalApplicationData maps to /data/user/0/com.companyname.mauiapptest/files locally in MAUI.

so I got the error solved by replacing this code: 
string createdFilePath = System.IO.Path.Combine(System.Environment.GetFolderPath(**System.Environment.SpecialFolder.Personal**), fileName);
with this one:
string createdFilePath = System.IO.Path.Combine(System.Environment.GetFolderPath(**Environment.SpecialFolder.LocalApplicationData**), fileName);
            
Now I got no errors.. thats good! : )

However, my App keep showing the loading bar (Unendlessly) with the text "Print Preview"!
I have tried both testing on the android amulator and the local Smartphone, I have tried also the default Printer and the PDF printer, no luck!
![image](https://github.com/dotnet/maui/assets/23061433/35943e9a-d0a9-48a8-a60e-ad5277a50e6b)

any advice is highly appreciated! 
Cheers!

**Reply by jaberlinz** [reply]
Update,
The error was related to the PDF File itself (probably because its in German!)
I replaced it with the PDF given in the example and it was printed smoothly!
Thanks McCarthy for your help : )

**Reply by LanceMcCarthy** [reply]
Aha! Nice catch :)

**Reply by ivanopaulon** [reply]
Hi! I have the same problem right now, but I need to understand why my PDF is not working.
I can preview and print it from Google Drive without problem, also in Windows I don't have problem printing it.
Do you have some suggestion? I create the file programmatically using a custom Font.
Some ideas?

**Reply by ivanopaulon** [reply]
i'm sorry, this is a file sample
[CONTO.PDF](https://github.com/user-attachments/files/16193745/CONTO.PDF)

**Reply by LanceMcCarthy** [reply]
@ivanopaulon When it comes to custom fonts, they usually need to be embedded in the document. This is a common misunderstanding even when using my company's document processing libraries, You can use our docs to learn more about the topic in general terms https://docs.telerik.com/devtools/document-processing/libraries/radpdfprocessing/concepts/fonts#embedded-fonts 

This should help you narrow down the same feature in the software that is being used to generate your PDF. Once the custom fonts are embedded, it is supposed to work anywhere.

**Reply by ivanopaulon** [reply]
Thanks, but my file has actually font embedded correctly, I'm using SharpPDF to create it.
![Screenshot 2024-07-12 160811](https://github.com/user-attachments/assets/79ad7640-90de-4207-9514-576940830f2b)

**Reply by LanceMcCarthy** [reply]
Okay, that's a good start. You stated that is works on Windows (WinUI), which platform is it not working on? This sounds like a good case for .NET MAUI issue.

**Reply by ivanopaulon** [reply]
Yes, debuggin my .Net Maui app in Windows it works, but on my Lenovo Tab P11 Pro, Android version 11 did not.
Also I can't understand why opening the file in Android with Acrobat Reader and printing it I had no problem,

**Reply by LanceMcCarthy** [reply]
This sounds like the Android native code needs a maui-android specialist. Unfortunately, I am not that. I did originally write and tweak the Windows code, then the rest of the platforms were added as time went on, but I've never had the time/ability to dig deeper in the Android and iOS side of things.

I have 2 suggestions to get some extra help with this:

1. Go to the .NET MAUI Discord server and go to the maui or android channels
   - If you're not a member yet, here's an invite link => https://discord.com/invite/HSuhTyG 
2. See if an Android specialist is available (someone like Peppers would be a perfect match)
3. Show the Android-native printing code and explain that it doesn't work when the PDF has custom fonts.

**Comment by PrabhuRout007** [comment]
Hey! @LanceMcCarthy what if I don't want to print the pdf I want to print the image which path I am passing in a string. I am getting the file but it seems it is not getting loaded in the previewPages but the data is present is canvas children element and some how print task is not working proper, I am sharing my code can you help me out with this.

public async void Print(Stream inputStream, string filepath, Window nativeWindow)
        {
            _filePath = filepath;
            this._fileName = Path.GetFileName(_filePath);
            this._window = nativeWindow;
            this._hWnd = WinRT.Interop.WindowNative.GetWindowHandle(_window);
            var fileType = File.ReadAllBytes(filepath);
            try
            {
                await PrintWithCanvas();
            }
            catch
            {
                Debug.WriteLine("No UI thread available, printing without Canvas.");
                await PrintWithoutCanvas();
            }
        }
        private async Task PrintWithCanvas()
        {
            await this.IncludeCanvas();
            this.RegisterForPrint();
            await PrintManagerInterop.ShowPrintUIForWindowAsync(_hWnd);
        }
        private async Task PrintWithoutCanvas()
        {
            this.RegisterForPrint();
            await PrintManagerInterop.ShowPrintUIForWindowAsync(_hWnd);
        }
        private void RegisterForPrint()
        {
            printDocument = new PrintDocument();
            printDocumentSource = printDocument.DocumentSource;
            printDocument.AddPages += AddPrintPages;
            printDocument.GetPreviewPage += PrintDoc_GetPreviewPage;
            printDocument.Paginate += PrintDoc_Paginate;
            printMan = PrintManagerInterop.GetForWindow(_hWnd);
            printMan.PrintTaskRequested += PrintTask_Requested;
        }
        private async void AddPrintPages(object sender, AddPagesEventArgs e)
        {
            try
            {
                await this.PrepareForPrint(0,1);

                ((PrintDocument)sender).AddPagesComplete();
            }
            catch
            {
                ((PrintDocument)sender).InvalidatePreview();
            }
        }
        private async Task PrepareForPrint(int p, int count)
        {
            imagePanel.Children.Clear();
            ApplicationLanguages.PrimaryLanguageOverride = CultureInfo.InvariantCulture.TwoLetterISOLanguageName;
            var imageCtrl = new Image();
            BitmapImage src = new BitmapImage(new Uri(_filePath, UriKind.Absolute));
            imageCtrl.Source = src;
            imageCtrl.Height = src.PixelHeight / DeviceDisplay.Current.MainDisplayInfo.Density;
            imageCtrl.Width = src.PixelWidth / DeviceDisplay.Current.MainDisplayInfo.Density;
            printDocument.AddPage(imageCtrl);
        }
        private async Task IncludeCanvas()
        {
            var imageCtrl = new Image();
            BitmapImage src = new BitmapImage(new Uri(_filePath, UriKind.Absolute));
            imageCtrl.Source = src;
            imageCtrl.Height = 816;
            imageCtrl.Width = 700;
            var pageCanvas = new Canvas
            {
                Width = 816,
                Height = 700,
                VerticalAlignment = VerticalAlignment.Top,
                HorizontalAlignment = HorizontalAlignment.Center,
                Background = new SolidColorBrush(Color.FromArgb(255, 255, 255, 255)),
                Margin = new Thickness(0, 0, 0, 0)
            };
            pageCanvas.Children.Add(imageCtrl);
            imagePanel.Children.Add(pageCanvas);
        }
        private void PrintTask_Requested(PrintManager sender, PrintTaskRequestedEventArgs args)
        {
            printTask = args.Request.CreatePrintTask(_fileName, sourceRequested => sourceRequested.SetSource(printDocumentSource));
            printTask.Completed += PrintTask_Completed;
        }

        private void PrintTask_Completed(PrintTask sender, PrintTaskCompletedEventArgs args)
        {
            printTask.Completed -= PrintTask_Completed;
            printMan.PrintTaskRequested -= PrintTask_Requested;
        }

        private async void PrintDoc_Paginate(object sender, PaginateEventArgs e)
        {
            var pageDescription = e.PrintTaskOptions.GetPageDescription((uint)e.CurrentPreviewPageNumber);

            marginWidth = pageDescription.PageSize.Width;
            marginHeight = pageDescription.PageSize.Height;

            for (var i = imagePanel.Children.Count - 1; i >= 0; i--)
            {
                var print = imagePanel.Children[i] as Canvas;

                if (print == null)
                    continue;

                print.Width = marginWidth;
                print.Height = marginHeight;

                this.printPreviewPages.TryAdd(i, print);
            }

        ((PrintDocument)sender).SetPreviewPageCount(_pageCount, PreviewPageCountType.Final);
        }
        private void PrintDoc_GetPreviewPage(object sender, GetPreviewPageEventArgs e)
        {
            imagePanel.Children.Remove(this.printPreviewPages[e.PageNumber - 1]);
            ((PrintDocument)sender).SetPreviewPage(e.PageNumber, this.printPreviewPages[e.PageNumber - 1]);
        }  
        
        passing the values to print parameter from here  
    private void PrintClickedEvent(object sender, EventArgs e)
    {
        try
        {
            var selectedImages = new List<String>();
            if (MainCanvasViewModel.CanvasImagePath != null && MainCanvasViewModel.IsCanvasHaveImage)
                selectedImages.Add(MainCanvasViewModel.CanvasImagePath);
            var filepath = selectedImages.FirstOrDefault();
            _printService = new PrintService();
            var openFile = System.IO.File.OpenRead(filepath);   
#if MACCATALYST
            _printService.PrintImageMac(openFile);
#elif WINDOWS
            _printService.Print(openFile, filepath, this.Window.Handler.PlatformView as Microsoft.UI.Xaml.Window);
#endif
        }
        catch (Exception ex)
        {
            LogHandler.LogError(ex);
        }
    }

**Reply by LanceMcCarthy** [reply]
Hi @PrabhuRout007 This workflow was specifically designed for PDF content bytes, so I can't promise it will work for other content types. That being said, I don't see why it couldn't work if the image byte content is correct and is loaded into an Image() control at runtime.

However, you may or may not see the preview correctly... so I recommend just sending it to the printer to double check if your issue is with the preview rendering or the entire setup.

Ultimately, this code originally came from Microsoft's PRinting API demo, you can refer back to that for more details on the implementation. It's more robust than this one, which was whittled down to support PDF layouts.

Here are links to two demos that implement printing:

- https://github.com/Microsoft/Windows-appsample-coloringbook
- https://github.com/microsoft/windows-universal-samples/tree/main/Samples/Printing

**Comment by PrabhuRout007** [comment]
@LanceMcCarthy Yes, although I am getting image bitmap and it is also added in the canvas as I debug I can see canvas has added the image as its children, when I print then also the print is happening smoothly, The issue is with the preview only, for first time when it is been called it has the the data but still not visible in preview although for the second time also I can the image is been displayed in canvas.

Here I had modified the code, if you found anything please let me know. Also it will help others who are facing issue in printing the image on Windows, All windows operations are working fine here except the previewPage functionality

    public class PrintService
    {
        private PageDescription currentPageDescription;
        private UIElement printPreviewPages;
        private Canvas imagePanel = new();
        private string _filePath;
        private Window _window;
        private string _fileName;
        private nint _hWnd;
        private PrintDocument printDocument;
        private IPrintDocumentSource printDocumentSource;
        private PrintManager printMan;
        private int _pageCount;
        private double marginWidth;
        private double marginHeight;
        private PrintTask printTask;
        private FrameworkElement firstPage;
        private PhotoSize photoSize;
        private Scaling photoScale;
        private const int DPI96 = 96;
#if MACCATALYST
        public async void PrintImageMac(Stream img)
        {
            try
            {
                var data = NSData.FromStream(img);
                var uiimage = UIImage.LoadFromData(data);

                var printer = UIPrintInteractionController.SharedPrintController;
                if (printer == null)
                {
                    await UIUtility.ShowMessage("Printer is not connected");
                }
                else
                {
                    var printInfo = UIPrintInfo.PrintInfo;
                    printInfo.OutputType = UIPrintInfoOutputType.General;
                    printer.PrintInfo = printInfo;
                    printer.PrintingItem = uiimage;

                    var handler = new UIPrintInteractionCompletionHandler((printInteractionController, completed, error) =>
                    {
                        if (completed)
                        {
                            UIUtility.ShowMessage("Printer Success");
                        }
                        else if (!completed && error != null)
                        {
                            UIUtility.ShowMessage("Print Failed");
                        }
                    });

                    CGRect frame = new CGRect();
                    frame.Size = uiimage.Size;
                    printer.Present(true, handler);
                }
            }
            catch (Exception ex)
            {
                LogHandler.LogError(ex);
            }
        }
#elif WINDOWS
        public async void Print(Stream inputStream, string filepath, Window nativeWindow)
        {
            _filePath = filepath;
            this._fileName = Path.GetFileName(_filePath);
            this._window = nativeWindow;
            this._hWnd = WinRT.Interop.WindowNative.GetWindowHandle(_window);
            var fileType = File.ReadAllBytes(filepath);
            try
            {
                await PrintWithCanvas();
            }
            catch(Exception ex)
            {
                LogHandler.LogError(ex);
            }
        }
        private async Task PrintWithCanvas()
        {
            try
            {
                this.RegisterForPrint();
                await PrintManagerInterop.ShowPrintUIForWindowAsync(_hWnd);
            }
            catch (Exception ex)
            {
                LogHandler.LogError(ex);
            }
        }

        private void RegisterForPrint()
        {
            try
            {
                printDocument = new PrintDocument();
                printDocumentSource = printDocument.DocumentSource;
                printDocument.Paginate += CreatePrintPreviewPages;
                printDocument.GetPreviewPage += PrintDoc_GetPreviewPage;

                printDocument.AddPages += AddPrintPages;
                printMan = PrintManagerInterop.GetForWindow(_hWnd);
                printMan.PrintTaskRequested += PrintTask_Requested;
            }
            catch (Exception ex)
            {
                LogHandler.LogError(ex);
            }
        }
        private async void AddPrintPages(object sender, AddPagesEventArgs e)
        {
            try
            {
                await this.PrepareForPrint();

                ((PrintDocument)sender).AddPagesComplete();
            }
            catch
            {
                ((PrintDocument)sender).InvalidatePreview();
            }
        }
        private async Task PrepareForPrint()
        {
            imagePanel.Children.Clear();
            ApplicationLanguages.PrimaryLanguageOverride = CultureInfo.InvariantCulture.TwoLetterISOLanguageName;
            var imageCtrl = new Image();
            BitmapImage src = new BitmapImage(new Uri(_filePath, UriKind.Absolute));
            imageCtrl.Source = src;

            imageCtrl.Height = currentPageDescription.ViewablePageSize.Height / DeviceDisplay.Current.MainDisplayInfo.Density;
            imageCtrl.Width = currentPageDescription.ViewablePageSize.Width / DeviceDisplay.Current.MainDisplayInfo.Density;
            imageCtrl.HorizontalAlignment = Microsoft.UI.Xaml.HorizontalAlignment.Center;
            imageCtrl.VerticalAlignment = Microsoft.UI.Xaml.VerticalAlignment.Center;
            printDocument.AddPage(imageCtrl);
        }

        private async void PrintTask_Requested(PrintManager sender, PrintTaskRequestedEventArgs args)
        {
            printTask = args.Request.CreatePrintTask(_fileName, sourceRequestedArgs =>
            {
                CustomizePrintTask(args, sourceRequestedArgs).GetAwaiter();
            });
            printTask.Completed += PrintTask_Completed1;
        }

        private void PrintTask_Completed1(PrintTask sender, PrintTaskCompletedEventArgs args)
        {
            printMan.PrintTaskRequested -= PrintTask_Requested;
        }
        private async Task CustomizePrintTask(PrintTaskRequestedEventArgs args, PrintTaskSourceRequestedArgs sourceRequestedArgs)
        {
            PrintTaskOptionDetails printDetailedOptions = PrintTaskOptionDetails.GetFromPrintTaskOptions(printTask.Options);
            printDetailedOptions.DisplayedOptions.Clear();
            printDetailedOptions.DisplayedOptions.Add(Windows.Graphics.Printing.StandardPrintTaskOptions.MediaSize);
            printDetailedOptions.DisplayedOptions.Add(Windows.Graphics.Printing.StandardPrintTaskOptions.Copies);
            printDetailedOptions.DisplayedOptions.Add(Windows.Graphics.Printing.StandardPrintTaskOptions.ColorMode);

            // Create a new list option.
            PrintCustomItemListOptionDetails photoSize = printDetailedOptions.CreateItemListOption("photoSize", "Photo Size");
            photoSize.AddItem("SizeFullPage", "Full Page");
            photoSize.AddItem("Size4x6", "4 x 6 in");
            photoSize.AddItem("Size5x7", "5 x 7 in");
            photoSize.AddItem("Size8x10", "8 x 10 in");
            printDetailedOptions.DisplayedOptions.Add("photoSize");
            printTask.Options.Orientation = PrintOrientation.Portrait;
            sourceRequestedArgs.SetSource(printDocumentSource);
        }
        private async void PrintDetailedOptionsOptionChanged(PrintTaskOptionDetails sender, PrintTaskOptionChangedEventArgs args)
        {
            if (args.OptionId == null)
            {
                return;
            }

            string optionId = args.OptionId.ToString();

            if (optionId == "photoSize")
            {
                IPrintOptionDetails photoSizeOption = sender.Options[optionId];
                string photoSizeValue = photoSizeOption.Value as string;

                if (!string.IsNullOrEmpty(photoSizeValue))
                {
                    switch (photoSizeValue)
                    {
                        case "SizeFullPage":
                            photoSize = PhotoSize.SizeFullPage;
                            break;
                        case "Size4x6":
                            photoSize = PhotoSize.Size4x6;
                            break;
                        case "Size5x7":
                            photoSize = PhotoSize.Size5x7;
                            break;
                        case "Size8x10":
                            photoSize = PhotoSize.Size8x10;
                            break;
                    }
                }
            }

            if (optionId == "scaling")
            {
                IPrintOptionDetails scalingOption = sender.Options[optionId];
                string scalingValue = scalingOption.Value as string;

                if (!string.IsNullOrEmpty(scalingValue))
                {
                    switch (scalingValue)
                    {
                        case "Crop":
                            photoScale = Scaling.Crop;
                            break;
                        case "ShrinkToFit":
                            photoScale = Scaling.ShrinkToFit;
                            break;
                    }
                }
            }
        }
        private async void CreatePrintPreviewPages(object sender, PaginateEventArgs e)
        {
            try
            {
                PrintDocument printDoc = (PrintDocument)sender;
                PageDescription pageDescription = new PageDescription();
                //PrintTaskOptionDetails printDetailedOptions = PrintTaskOptionDetails.GetFromPrintTaskOptions(e.PrintTaskOptions);
                PrintPageDescription printPageDescription = e.PrintTaskOptions.GetPageDescription(0);
                pageDescription.Margin.Width = Math.Max(
                    printPageDescription.ImageableRect.Left,
                    printPageDescription.ImageableRect.Right - printPageDescription.PageSize.Width);

                pageDescription.Margin.Height = Math.Max(
                    printPageDescription.ImageableRect.Top,
                    printPageDescription.ImageableRect.Bottom - printPageDescription.PageSize.Height);

                pageDescription.ViewablePageSize.Width = printPageDescription.PageSize.Width - pageDescription.Margin.Width * 2;
                pageDescription.ViewablePageSize.Height = printPageDescription.PageSize.Height - pageDescription.Margin.Height * 2;

                switch (photoSize)
                {
                    case PhotoSize.Size4x6:
                        pageDescription.PictureViewSize.Width = 4 * DPI96;
                        pageDescription.PictureViewSize.Height = 6 * DPI96;
                        break;
                    case PhotoSize.Size5x7:
                        pageDescription.PictureViewSize.Width = 5 * DPI96;
                        pageDescription.PictureViewSize.Height = 7 * DPI96;
                        break;
                    case PhotoSize.Size8x10:
                        pageDescription.PictureViewSize.Width = 8 * DPI96;
                        pageDescription.PictureViewSize.Height = 10 * DPI96;
                        break;
                    case PhotoSize.SizeFullPage:
                        pageDescription.PictureViewSize.Width = pageDescription.ViewablePageSize.Width;
                        pageDescription.PictureViewSize.Height = pageDescription.ViewablePageSize.Height;
                        break;
                }
                if ((pageDescription.ViewablePageSize.Width > pageDescription.ViewablePageSize.Height) && (photoSize != PhotoSize.SizeFullPage))
                {
                    var swap = pageDescription.PictureViewSize.Width;
                    pageDescription.PictureViewSize.Width = pageDescription.PictureViewSize.Height;
                    pageDescription.PictureViewSize.Height = swap;
                }
                pageDescription.IsContentCropped = photoScale == Scaling.Crop;
                currentPageDescription = pageDescription;
                printDoc.SetPreviewPageCount(1, PreviewPageCountType.Intermediate);
            }
            catch (Exception ex)
            {
                LogHandler.LogError(ex);
            }
            
        }
        private async void PrintDoc_GetPreviewPage(object sender, GetPreviewPageEventArgs e)
        {
            int pageNumber = e.PageNumber;
            UIElement page;
            page = await GeneratePageAsync(1, currentPageDescription);
            printPreviewPages = page;
            PrintDocument printDoc = (PrintDocument)sender;
            printDoc.SetPreviewPage(pageNumber, page);

        }
        private async Task<UIElement> GeneratePageAsync(int photoNumber, PageDescription pageDescription)
        {
            Canvas viewablePage = new Canvas()
            {
                Width = pageDescription.ViewablePageSize.Width,
                Height = pageDescription.ViewablePageSize.Height
            };

            var imageCtrl = new Image();
            BitmapImage src = new BitmapImage(new Uri(_filePath, UriKind.Absolute));
            imageCtrl.Source = src;

            imageCtrl.Height = currentPageDescription.ViewablePageSize.Height / DeviceDisplay.Current.MainDisplayInfo.Density;
            imageCtrl.Width = currentPageDescription.ViewablePageSize.Width / DeviceDisplay.Current.MainDisplayInfo.Density;
            imageCtrl.HorizontalAlignment = Microsoft.UI.Xaml.HorizontalAlignment.Center;
            imageCtrl.VerticalAlignment = Microsoft.UI.Xaml.VerticalAlignment.Center;

            viewablePage.Children.Add(imageCtrl);
            return viewablePage;
        }
        private static void Swap<T>(ref T v1, ref T v2)
        {
            T swap = v1;
            v1 = v2;
            v2 = swap;
        }
#endif
    }
}
public class PageDescription : IEquatable<PageDescription>
{
    public Size Margin;
    public Size PageSize;
    public Size ViewablePageSize;
    public Size PictureViewSize;
    public bool IsContentCropped;

    public bool Equals(PageDescription other)
    {
        // Detect if PageSize changed
        bool equal = (Math.Abs(PageSize.Width - other.PageSize.Width) < double.Epsilon) &&
            (Math.Abs(PageSize.Height - other.PageSize.Height) < double.Epsilon);

        // Detect if ViewablePageSize changed
        if (equal)
        {
            equal = (Math.Abs(ViewablePageSize.Width - other.ViewablePageSize.Width) < double.Epsilon) &&
                (Math.Abs(ViewablePageSize.Height - other.ViewablePageSize.Height) < double.Epsilon);
        }
        // Detect if PictureViewSize changed
        if (equal)
        {
            equal = (Math.Abs(PictureViewSize.Width - other.PictureViewSize.Width) < double.Epsilon) &&
                (Math.Abs(PictureViewSize.Height - other.PictureViewSize.Height) < double.Epsilon);
        }
        // Detect if cropping changed
        if (equal)
        {
            equal = IsContentCropped == other.IsContentCropped;
        }
        return equal;
    }
}
public enum PhotoSize : byte
{
    SizeFullPage,
    Size4x6,
    Size5x7,
    Size8x10
}
public enum Scaling : byte
{
    ShrinkToFit,
    Crop
}

**Reply by LanceMcCarthy** [reply]
Try dispatching the Preview-rendering cycle to the UI thread. I removed that to keep my example simple. You can also go to the original Microsoft demo to see the full implementation.

**Reply by Harrynho** [reply]
Hey Lance, Is there a chance your solution would work on a Maui Blazor Hybrid web app? I'm trying to develop a POS for windows using maui and blazor components, i added your PrinterHelper.Windows.cs, but since the razor components are on the (Blazor project) Shared project, i cannot use the "this.Window.Handler.PlatformView as Microsoft.UI.Xaml.Window" parameter for the helper.Print function. Have you try printing on a maui blazor hybrid solution? Thanks in advance.


-------------------------------------SOLUTION---------------------
Instead of "this.Window.Handler.PlatformView as Microsoft.UI.Xaml.Window", you can use "Application.Current.Windows[0].Handler.PlatformView as Microsoft.UI.Xaml.Window". For anyone using maui blazor hybrid, run direct to "PrintWithoutCanvas". You are a genius Lance. Now i'm trying to do a silent print...

**Reply by LanceMcCarthy** [reply]
Bingo, they are essentially the same object :)

**Comment by ahmedyaseen3010** [comment]
> how to print pdf?
Here's a work around solution for POS devices 
https://github.com/ahmedyaseen3010/PDF-Printing-on-POS-for-Maui.net-
